compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/ClosureDesugar.java

```diff
@@ -808,12 +981,90 @@ public void visit(BLangIndexBasedAccess indexAccessExpr) {
         result = desugar.rewriteExpr(indexAccessExpr);
     }
 
+    public BPackageSymbol getPackageSymbol(BSymbol symbol) {
+        if (symbol.owner.getKind() == SymbolKind.PACKAGE) {
+            return (BPackageSymbol) symbol.owner;
+        }
+        return getPackageSymbol(symbol.owner);
+    }
+
     @Override
-    public void visit(BLangInvocation iExpr) {
-        iExpr.expr = rewriteExpr(iExpr.expr);
-        iExpr.requiredArgs = rewriteExprs(iExpr.requiredArgs);
-        iExpr.restArgs = rewriteExprs(iExpr.restArgs);
-        result = iExpr;
+    public void visit(BLangInvocation invocation) {
+        rewriteInvocationExpr(invocation);
+    }
+
+    public void rewriteInvocationExpr(BLangInvocation invocation) {
```

Comment:
Isn't this the logic that adds invokes the parameter's default value closure if the user has not provided an argument for that param? IMO, that should go in the desugar. The ClosureDesugar basically does the desugaring required for closures to work. Invoking the closure does not belong in this class IMO.

Reply:
For closures, we inject the parameters(closureMap) in the closure desugar. phase. Due to that, I think we can't move this to desugar because without knowing the parameters of that closure we can't create the invocation.

Reply:
Didn't we discuss to move that entire logic to desugar or a new kind of desugar like annotation or service desugar?

IMO, as far as the closure desugar is concerned, by the time we get here

```ballerina
int x = 123;

function fn(int i = x) {
}

public function main() {
    fn();
}
```

can be equivalent to something like

```ballerina
int x = 123;

function fn(int i = x) {
}

public function main() {
    function () returns int \$fn\$param\$i = function () returns int {
        return x;
    };
    fn(\$fn\$param\$i());
}
```

Reply:
Add new phase for this.